adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java

```diff
@@ -482,4 +486,43 @@ public FareRule getFareRule(final String fareId, final String routeId, final Str
         return fareRuleCollection.get(FareRule.getFareRuleMappingKey(fareId, routeId, originId, destinationId,
                 containsId));
     }
+
+    /**
+     * Add a shape point representing a row from shapes.txt to this. Return the entity added to the repository if not
+     * null, otherwise returns null.
+     *
+     * @param newShapePoint the internal representation of a row from shapes.txt to be added to the repository.
+     * @return the entity added to the repository if not null, otherwise returns null.
+     * @throws IllegalArgumentException if the shape point passed as argument is null
+     */
+    @Override
+    public ShapePoint addShapePoint(final ShapePoint newShapePoint) throws IllegalArgumentException {
+        if (newShapePoint != null) {
+            final String shapeId = newShapePoint.getShapeId();
+            if (shapePerId.containsKey(shapeId)) {
+                if (shapePerId.get(shapeId).contains(newShapePoint)) {
```

Comment:
!! be wary of how `contains` behave. I suppose it does a filed by field comparison but that should be checked. Also it seems we are missing a test where trying to add twice two different ShapePoint *instances* that specify the same id and data.

Reply:
This `.contains()` is on the `ArrayList`, right? So the time complexity here is O(n) to find out if the new shape point exists for this shape_id. Given that shapes.txt can be huge, this is very expensive.

I'd strongly suggest a more efficient solution with `TreeMap` instead:
https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html

That will give you O(log n) time complexity for `.contains()`, and you can still get sorted shape points back.

Reply:
And as implied above, we should store these points already sorted in the repo (see my comment later for adding a test for this).